### PR TITLE
Force download doesn't require to do a head request

### LINKED ISSUE #14837: Querying datasources over HTTP fails if the server rejects HEAD requests.
### What happens?

Using any `read_xxx` method on a resources that blocks `HEAD` requests with a non 2xx status code fails with an IO-error

```
IO Error: Unable to connect to URL "https://api.spring.io/projects/spring-boot/generations": 401 (Unauthorized)
```

It's kinda unexpected, as the HTTP logging shows both a `HEAD` and a `GET`, the latter being successful. 

```
HTTP Request:
	HEAD /projects/spring-boot/generations
	Accept: */*
	Host: api.spring.io
	User-Agent: cpp-httplib/0.14.3

HTTP Response:
	401 Unauthorized 
	alt-svc: h3=":443"; ma=86400
	Cache-Control: no-cache, no-store, max-age=0, must-revalidate
	CF-Cache-Status: DYNAMIC
	CF-RAY: 8e28db679e65c7ec-DUS
	Connection: keep-alive
	Content-Length: 0
	Date: Thu, 14 Nov 2024 17:52:45 GMT
	expires: 0
	pragma: no-cache
	Server: cloudflare
	Set-Cookie: JSESSIONID=51FE8280FD415EF841542D9519B9E726; Path=/; Secure; HttpOnly
	Set-Cookie: __cf_bm=wSxGWuyIB3Yt_hWYDTISeQnIq_XLf092PgnIZyc7tzI-1731606765-1.0.1.1-fAoBhY_Ch_GtXEXxlWSIn5PwO0aYptSSKEIVeR0OUNqJNkqhimQuR4FzeACExsWMNGd_nFzhp0Kd89DNKwZfmQ; path=/; expires=Thu, 14-Nov-24 18:22:45 GMT; domain=.spring.io; HttpOnly; Secure; SameSite=None
	strict-transport-security: max-age=15552000; includeSubDomains
	vary: Origin
	vary: Access-Control-Request-Method
	vary: Access-Control-Request-Headers
	www-authenticate: Basic realm="Realm"
	x-content-type-options: nosniff
	x-frame-options: DENY
	x-xss-protection: 0


HTTP Request:
	GET /projects/spring-boot/generations
	Accept: */*
	Host: api.spring.io
	Range: bytes=0-1
	User-Agent: cpp-httplib/0.14.3

HTTP Response:
	200 OK 
	alt-svc: h3=":443"; ma=86400
	Cache-Control: no-cache, no-store, max-age=0, must-revalidate
	CF-Cache-Status: DYNAMIC
	CF-RAY: 8e28db69992cc7ec-DUS
	Connection: keep-alive
	Content-Type: application/hal+json
	Date: Thu, 14 Nov 2024 17:52:45 GMT
	expires: 0
	pragma: no-cache
	Server: cloudflare
	Set-Cookie: __cf_bm=y_2x58.2Qcv5d5n1ARIc_pQYbOmjOrsH6VUE52oPe4s-1731606765-1.0.1.1-g9OpXf3bc0.QfzAPZyHqOAsqlJVjlxdht_Z40CngN2hnFeZQP8K8WH2mMuOosXEed81hOAJal6AXnSlfIh9hZw; path=/; expires=Thu, 14-Nov-24 18:22:45 GMT; domain=.spring.io; HttpOnly; Secure; SameSite=None
	strict-transport-security: max-age=15552000; includeSubDomains
	Transfer-Encoding: chunked
	vary: Origin
	vary: Access-Control-Request-Method
	vary: Access-Control-Request-Headers
	x-content-type-options: nosniff
	x-frame-options: DENY
	x-xss-protection: 0
```

### To Reproduce

```
duckdb -s "SET enable_http_logging=true" -s "SELECT * FROM read_json('https://api.spring.io/projects/spring-boot/generations')"
```

### OS:

macOS

### DuckDB Version:

v1.1.3 19864453f7

### DuckDB Client:

DuckDB CLI

### Hardware:

_No response_

### Full Name:

Michael Simons

### Affiliation:

Neo4j, Inc.

### What is the latest build you tested with? If possible, we recommend testing with the latest nightly build.

I have tested with a stable release

### Did you include all relevant data sets for reproducing the issue?

Yes

### Did you include all code required to reproduce the issue?

- [X] Yes, I have

### Did you include all relevant configuration (e.g., CPU architecture, Python version, Linux distribution) to reproduce the issue?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by pdet**
Fix: #14837

### CODE REVIEW

**Comment by pdet** [comment]
We could also just try the force download automatically if the head request fails.
I don't have a strong opinion/preference about it, what you think @samansmink ?

**Comment by samansmink** [comment]
> We could also just try the force download automatically if the head request fails.

My preference would be to throw an error that hints switching on full file downloads instead of silently falling back to it because it can significantly change query performance.

File: extension/httpfs/httpfs.cpp [hunk 1]
```diff
@@ -700,7 +700,7 @@ void HTTPFileHandle::Initialize(optional_ptr<FileOpener> opener) {
 			}
 			length = 0;
 			return;
-		} else {
+		} else if (!http_params.force_download){
```
Comment (samansmink) [inline extension/httpfs/httpfs.cpp:703]: just wondering: should we even do a head request for the full file download?

**Comment by pdet** [comment]
Great suggestions @samansmink, added both! :-)

**Comment by Mytherin** [comment]
Thanks for the extra info - after some more discussion we think it's a good idea to accept `200` for range requests and fallback to full file downloads in that scenario. Servers that don't support this likely are not serving large files anyway and the user is most likely not in control of the server in either case, so requiring the user to set an option to interact with that server when other tools work with it out-of-the-box would just add unnecessary annoyance.

**Comment by carlopi** [comment]
I think IF response is 200, we will get the full file right there, and if response is 206, than it's the requested ranges.

This should simplify things.

One quirk though is that if we get the file on a range request whose answer is 200, we should save it somewhere, instead of downloading it again.

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**